### ISSUE TITLE
Add Hurl to official Debian repository

### ISSUE THREAD

**Issue by fabricereix** [Orange-OpenSource/hurl#366]

**Comment by lepapareil** [comment]
21/12/15 > Mail sent to `owner@bugs.debian.org`

**Comment by jcamiel** [comment]
Hi @stappersg  

It's just a note to ourself to keep an update on trying to add Hurl to official Debian repo.

We've 0 experience in this domain and struggle a bit to know where to start.
We've send a first mail to <submit@bugs.debian.org>, as described in <https://wiki.debian.org/DebianMentorsFaq#:~:text=File%20an%20ITP%20(intent%20to%20package)%20bug%20using%20reportbug> to make a official request to propose Hurl.
Do you think you can help us ? Or share any experience/tips to guide us ?
Thanks, 

Jc

**Comment by stappersg** [comment]
On Wed, Dec 15, 2021 at 10:17:28AM -0800, jcamiel wrote:
> We've send a first mail to < submit AT bugs.debian.org >

OK, that is the correct email address for reporting an "bug".

Meanwhile should there be bug report number.  What is it?


> Do you think you can help us ?

I doing so by asking good questions.


> Or share any experience/tips to guide us ?

Work in progress, currently low priority ...

**Comment by lepapareil** [comment]
Hi @stappersg, here is the bug report number https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1001779.

**Comment by lepapareil** [comment]
Hi @stappersg, we probably need some help here...

Like @jcamiel said it's our first experience attempting to create a Debian package, so i wonder if we have forgotten something to move [it](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1001779) forward since our bug seems to be on hold for almost 3 months ?

Thks for your help :)

**Comment by stappersg** [comment]
On Thu, Mar 24, 2022 at 12:23:16AM -0700, lepapareil wrote:
> ... our bug seems to be on hold for almost 3 months ?

No, nothing is on hold.  An request for packaging
is a request for packaging.


> Thks for your help :)

Expressing expectations is fine,
insisting on ones own expectation doesn't help.


FWIW, I only partly argee with "somebody else should do it",
thing I will do is sponsoring uploads to Debian.

On expressing expectations:  I love to see messages like

  At $URL is hurl dependency foo packaged. What is,
  beside review time, needed to get it uploaded to Debian?



Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by lepapareil** [comment]
thks @stappersg

**Comment by jcamiel** [comment]
Doing small progress here. We can start by looking at packaging `hurl-core`. With [cargo-debstatus](https://crates.io/crates/cargo-debstatus), we can look which crate is available on Debian Unstable:

```shell
$ cargo debstatus --package hurl_core
 🔴 hurl_core v6.1.0-SNAPSHOT (/Users/jc/Documents/Dev/hurl/packages/hurl_core)
 ⌛ ├── colored v3.0.0 (outdated, 2.1.0 in debian)
 🔴 ├── libxml v0.3.3
    │   └── libc v0.2.169 (in debian)
    │   [build-dependencies]
    │   └── pkg-config v0.3.31 (in debian)
    └── regex v1.11.1 (in debian)
```

List of Rust packets available on Debian unstable (sid): <https://packages.debian.org/sid/rust/>.

Check that a `foo` crate is  available : <https://packages.debian.org/sid/rust/librust-foo-dev>

=> We can see that the crate `libxml` is not in Debian Unstable. A really good first step would be to packetize `libxml` crate, this is a mandatory step for `hurl_core` to be available as a Debian packet.

We can do the same with `libxml`:

```shell
$ cargo debstatus
 🔴 libxml v0.3.3 (/Users/jc/Documents/Dev/rust-libxml)
    └── libc v0.2.169 (in debian)
    [build-dependencies]
    └── pkg-config v0.3.31 (in debian)
    [dev-dependencies]
    ├── criterion v0.5.1 (in debian)
    └── rayon v1.10.0 (in debian)
```

Every dependency crate of `libxml` crate is available on Debian unstable. We can try to make a first packet for it.

## Regarding `hurl`

This is the list of crate that are not available in Debin Unstable:

```shell
cargo debstatus --package hurl
 🔴 hurl v6.1.0-SNAPSHOT (/Users/jc/Documents/Dev/hurl/packages/hurl)
 🔴 ├── hurl_core v6.1.0-SNAPSHOT (/Users/jc/Documents/Dev/hurl/packages/hurl_core)
 ⌛ │   ├── colored v3.0.0 (outdated, 2.1.0 in debian)
 🔴 │   ├── libxml v0.3.3
    │   │   └── libc v0.2.169 (in debian)
    │   │   [build-dependencies]
    │   │   └── pkg-config v0.3.31 (in debian)
    │   └── regex v1.11.1 (in debian)
    ├── base64 v0.22.1 (in debian)
 ⌛ ├── brotli v7.0.0 (outdated, 6.0.0 in debian)
    │   ├── alloc-no-stdlib v2.0.4 (in debian)
    │   ├── alloc-stdlib v0.2.2 (in debian)
    │   └── brotli-decompressor v4.0.2 (4.0.1 in debian)
    ├── chrono v0.4.39 (in debian)
    ├── clap v4.5.29 (4.5.23 in debian)
    ├── curl-sys v0.4.79+curl-8.12.0 (0.4.67 in debian)
    ├── curl v0.4.47 (0.4.44 in debian)
    ├── encoding v0.2.33 (in debian)
    ├── glob v0.3.2 (in debian)
    ├── hex v0.4.3 (in debian)
    ├── lazy_static v1.5.0 (in debian)
 ⌛ ├── libflate v2.1.0 (outdated, 0.1.25 in debian)
    │   ├── adler32 v1.2.0 (in debian)
    │   ├── core2 v0.4.0 (in debian)
    │   ├── crc32fast v1.4.2 (in debian)
    │   ├── dary_heap v0.3.7 (0.3.6 in debian)
 ⌛ │   └── libflate_lz77 v2.1.0 (outdated, 1.1.0 in debian)
    │       ├── core2 v0.4.0 (in debian)
    │       ├── hashbrown v0.14.5 (in debian)
    │       └── rle-decode-fast v1.0.3 (in debian)
 🔴 ├── libxml v0.3.3
    │   └── libc v0.2.169 (in debian)
    │   [build-dependencies]
    │   └── pkg-config v0.3.31 (in debian)
    ├── md5 v0.7.0 (in debian)
    ├── percent-encoding v2.3.1 (in debian)
    ├── regex v1.11.1 (in debian)
    ├── serde v1.0.217 (in debian)
    ├── serde_json v1.0.138 (1.0.134 in debian)
    ├── sha2 v0.10.8 (in debian)
    ├── similar v2.7.0 (2.6.0 in debian)
    ├── terminal_size v0.4.1 (in debian)
 ⌛ ├── termion v4.0.3 (outdated, 1.5.6 in debian)
    │   ├── libc v0.2.169 (in debian)
    │   └── numtoa v0.2.4 (in debian)
    ├── url v2.5.4 (2.5.2 in debian)
    ├── uuid v1.13.1 (1.10.0 in debian)
    └── xml-rs v0.8.25 (0.8.19 in debian)
    [build-dependencies]
    └── cc v1.2.13 (1.1.14 in debian)
```

Not regarding the version number, only `hurl_core` and `libxml` are not in Debian Unstable.

**Comment by wetneb** [comment]
I've started packaging libxml.

**Comment by wetneb** [comment]
I have made a [merge request](https://salsa.debian.org/rust-team/debcargo-conf/-/merge_requests/852) for packaging libxml. Once this gets merged, someone needs to do the upload, after which it will take a while for the package to go through the NEW queue. Only after that we'll be able to do the same dance for hurl

**Comment by stappersg** [comment]
On Fri, Feb 28, 2025 at 09:50:43AM -0800, Antonin Delpeuch wrote:
( https://github.com/Orange-OpenSource/hurl/issues/366#issuecomment-2691228495 )
> wetneb left a comment (Orange-OpenSource/hurl#366)
> 
> I have made a [merge
> request](https://salsa.debian.org/rust-team/debcargo-conf/-/merge_requests/852)
> for packaging libxml. Once this gets merged, someone needs to do
> the upload, after which it will take a while for the package to go
> through the NEW queue. Only after that we'll be able to do the same
> dance for hurl
> 

Hello Antonin,
Hello all others who see this email,
 

What is the further plan?

Yes, that is a honest question.
Thing is that I would like to help, but I can't help.
That is not completely true, I'm writing this email.


In the Debian changelog (
https://salsa.debian.org/rust-team/debcargo-conf/-/merge_requests/852/diffs?file=10828e0258fa3697133c24f6d970e59f69569f89#diff-content-10828e0258fa3697133c24f6d970e59f69569f89
)
do I miss "Closes #NNNNNN" ( #NNNNNN the Intent To Package bugreport )

I could be wrong, that Debian Rust packages in "the huge debcargo repo"
don't need ITP bugreports. I don't know, it is the reason why I can't
help.  Hence my "What is the further plan?".

I do hope for a response like "I'm following the process described
at ...some_URL...".

Due  ***@***.*** in the CC, I hope
for a "Do follow procedure described at ....URL...."


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by wetneb** [comment]
Hi @stappersg,

Quoting https://wiki.debian.org/Teams/RustPackaging:
> Binary packages (such as applications) should be maintained also with the rust team if they are available as crates. Please file an ITP for those, not for regular crates. Examples: lsd, exa, bat ... 

Since my merge request is only packaging libxml, a crate which doesn't expose a binary on its own, I didn't file an ITP.

**Comment by jcamiel** [comment]
Hi @stappersg 

As one of the maintainer of this package, I'm trying really hard to understand your messages:

> Thing is that I would like to help, but I can't help.

English is not my primarily language so maybe there is some misunderstanding. Any help on packaging Hurl for Debian is welcome. But if you think you can't help us, what's the point of it, besides confusing us?

> I do hope for a response like "I'm following the process described
at ...some_URL...".: 

I do not understand this sentence at all, unless we're being trolled. You seem to know a lot about packaging for Debian and willing to help us, but, at this point, it's counter productive.

**Comment by stappersg** [comment]
On Fri, Feb 28, 2025 at 12:35:55PM -0800, Antonin Delpeuch wrote:
> stappers wrote:
> > What is the plan?
> 
> Quoting https://wiki.debian.org/Teams/RustPackaging:
> > Binary packages (such as applications) should be maintained also
> > with the rust team if they are available as crates. Please file an
> > ITP for those, not for regular crates. Examples: lsd, exa, bat ...
> 
> Since my merge request is only packaging libxml, a crate which doesn't
> expose a binary on its own, I didn't file an ITP.
> 

I see

Thanks


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by jcamiel** [comment]
Hi @wetneb 

Any help is welcome on this issue thanks a lot. We've packaged Hurl for a lot of platform and I've to confess that we're struggling with Debian. Something that is not clear in my mind: let's say we manage to get all the necessary crate package as Debian package for Hurl. How do you deal with version? For instance, current Hurl 6.0.0 version depends on `colored 3.0.0` but the Debian package is only at `2.1.0`.

How do we deal with it? Do we need to package a less recent version of Hurl (let's say 4.0.0) for which each crate is packaged at a sufficient version?

Thanks a lot!

**Comment by nc7s** [comment]
Geert: [ITP #1001779](https://bugs.debian.org/1001779) has been filed for hurl (converted from RFP, strictly speaking; also the conversation history is a bit messy). Maytham Alsudany currently owns it, maytham on IRC.

@jcamiel: No, we can relax the version requirements and patch relevant usage accordingly at our end.

**Comment by jcamiel** [comment]
@nc7s thanks for the information, it's still a little unclear how this will works (for instance, if Hurl needs a major breaking dependency 2.x while there is only a 1.x in Debian).

On our side (upstream maintainers) we're going to remove, in the next Hurl version, two dependencies (`colored` and `termion`) to reduce the crates we depends on.

**Comment by nc7s** [comment]
Taking your example of colored: we can patch hurl to downgrade to colored 2.1.0 (in Debian) from 3.0.0 (upstream).

There's no need from us, and I hope the decision is not because of us, to remove dependencies to ease packaging, at the cost of functionalities. If some dependencies are really complicated to package, we would do that ourselves.

**Comment by Maytha8** [comment]
> Geert: [ITP #1001779](https://bugs.debian.org/1001779) has been filed for hurl (the conversation history is a bit messy). Maytham Alsudany currently owns it, maytham on IRC.

👋 Hello! rust-hurl & rust-hurl-core are now in the NEW queue.

**Comment by jcamiel** [comment]
Thank you very much for this good news @Maytha8 . Just two questions :

- will the binary package be named `hurl`, available with `apt install hurl`?
- there is a helper binary with `hurl`, called `hurlfmt` that allows to export Hurl files to various formats. With other platforms, we usually provide the two binaries in the same packet. Will the Hurl Debian packet provide `hurlfmt` alongside `hurl`?

Thanks a lot!

**Comment by Maytha8** [comment]
On Thu, 2025-03-06 at 00:49 -0800, Jean-Christophe Amiel wrote:
> Thank you very much for this good news @Maytha8 . Just two questions :
>  * will the binary package be named hurl, available with apt install
> hurl?

Yes.

>  * there is a helper binary with hurl, called hurlfmt that allows to
> export Hurl files to various formats. With other platforms, we usually
> provide the two binaries in the same packet. Will the Hurl Debian
> packet provide hurlfmt alongside hurl?
> Thanks a lot!

Since hurlfmt is in a separate crate, it will need to be packaged
separately (which I will do soon when I time). I can have the hurl
package recommend hurlfmt (not depend, since hurlfmt is not essential to
the operation of hurl), and hurlfmt will be installed alongside hurl
unless the user explicitly passes --no-install-recommends to apt.

**Comment by jcamiel** [comment]
Hi @nc7s @Maytha8 I've seen that the Hurl Sid packet <https://packages.debian.org/sid/hurl> has been upgraded to 6.1.1, which is cool.Two quick questions:

- what's the process for a packet to go from sid to testing? Is it done automatically after a few months in unstable or do someone has to do something?
- we're going to release a 7.0.0 version at the end of the month, do we need to advertise this to someone for the Hurl packet to be upgraded?

Thanks and sorry if there are "naive" questions!

**Comment by Maytha8** [comment]
On Sat, 2025-07-19 at 04:55 -0700, Jean-Christophe Amiel wrote:
>  * what's the process for a packet to go from sid to testing? Is it
> done automatically after a few months in unstable or do someone has to
> do something?

Packages will go from unstable to testing once they have been in unstable for a
few days, build successfully and pass tests on all of Debian's release
architectures, have no release-critical bugs, and all dependencies are
satisfiable. More detailed info can be found [here](https://www.debian.org/devel/testing).

Currently, a hard freeze is in effect as Debian 13 (Trixie) is about to
be released, so nearly nothing is able to automatically migrate.

>  * we're going to release a 7.0.0 version at the end of the month, do
> we need to advertise this to someone for the Hurl packet to be
> upgraded?

I'm already watching releases so will recieve emails for that. Other
than that, it just depends on whether the maintainer notices that there
is a new version available. It helps that the [Package Tracking
System](https://tracker.debian.org) displays a warning when there is a
newer version available.

**Comment by jcamiel** [comment]
Thanks @Maytha8 very clear!

**Comment by jcamiel** [comment]
Updates: @fabricereix @lepapareil Hurl packets are now in testing (forky) => <https://packages.debian.org/forky/hurl> <https://packages.debian.org/source/forky/rust-hurl> etc..., it's happening!

**Comment by jcamiel** [comment]
Tracker's URL <https://tracker.debian.org/pkg/rust-hurl>

**Comment by jgarte** [comment]
Hi, should this issue be closed now since it looks like hurl was added to Debian? @jcamiel

**Comment by stappersg** [comment]
> Hi, should this issue be closed now since it looks like hurl was added to Debian?

https://packages.debian.org/source/testing/rust-hurl shows `hurl` is in Debian.

Yes, this issue should be closed.

**Comment by jcamiel** [comment]
Hi all, I'll close the issue when we'll be able to run `apt install hurl` on stable Debian (and we'll update Hurl installation instruction)
